drivers/ethos/ethos.c

```diff
@@ -0,0 +1,322 @@
+/*
+ * Copyright (C) 2015 Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     driver_ethos
+ * @{
+ *
+ * @file
+ * @brief       Implementation of a simple ethernet-over-serial driver
+ *
+ * @author      Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * @}
+ */
+
+#include <assert.h>
+#include <errno.h>
+#include <string.h>
+
+#include "random.h"
+#include "ethos.h"
+#include "periph/uart.h"
+#include "tsrb.h"
+
+#include "net/netdev2.h"
+#include "net/netdev2_eth.h"
+#include "net/eui64.h"
+#include "net/ethernet.h"
+
+#ifdef USE_ETHOS_FOR_STDIO
+#include "uart_stdio.h"
+#endif
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+static void _get_mac_addr(netdev2_t *dev, uint8_t* buf);
+static void ethos_isr(void *arg, char c);
+const static netdev2_driver_t netdev2_driver_ethos;
+
+static const uint8_t _esc_esc[] = {ETHOS_ESC_CHAR, (ETHOS_ESC_CHAR ^ 0x20)};
+static const uint8_t _esc_delim[] = {ETHOS_ESC_CHAR, (ETHOS_FRAME_DELIMITER ^ 0x20)};
+
+
+void ethos_setup(ethos_t *dev, uart_t uart, uint32_t baudrate, uint8_t *buf, size_t bufsize)
+{
+    dev->netdev.driver = &netdev2_driver_ethos;
+    dev->uart = uart;
+    dev->state = WAIT_FRAMESTART;
+    dev->framesize = 0;
+    dev->frametype = 0;
+    dev->last_framesize = 0;
+
+    tsrb_init(&dev->inbuf, (char*)buf, bufsize);
+    mutex_init(&dev->out_mutex);
+
+    uint32_t a = genrand_uint32();
```

Comment:
I already hate the fact, that `netdev2_tap` has more or less random MAC addresses, but at least they are somewhat stable as long as you don't rebuild the TAP. Can't you use the CPUID to generate the MAC? It would make debugging and testing with these devices way easier ;-).